### PR TITLE
TNPG updated to work with FCM push v1.

### PR DESCRIPTION
**Pull request by or-else**
TNPG updated
FCM push refactored a bit to use the new error handling. 
Common code moved to a separate package.

All pushes seem to work, including voip. But voip pushes arrive as normal pushes as opposite to PushKit despite the fact that they are sent as voip.

### CODE REVIEW

**Comment by or-else** [comment]
The new PushGW is also in `fcmv1` branch. It needs work though: I converted it to use v1 only, but it still needs to support v0 while we transition. i'll add v0 back.

File: server/push/fcm/payload.go [hunk 1]
```diff
@@ -597,10 +196,11 @@ func PrepareV1Notifications(rcpt *push.Receipt, config *configType) []*fcmv1.Mes
 		// We don't know the platform of the receiver, must provide payload for all platforms.
 		msg.Android = androidNotificationConfig(rcpt.Payload.What, topic, userData, config)
 		msg.Apns = apnsNotificationConfig(rcpt.Payload.What, topic, userData, 0, config)
+		// TODO: add webpush payload.
 		messages = append(messages, &msg)
 	}
 
-	return messages
+	return messages, uids
```
Comment (aforge) [inline server/push/fcm/payload.go:203]: Is `len(messages) == len(uids)` an invariant here?
If it is, channel messages (line 200) breaks it.

File: server/push/tnpg/push_tnpg.go [hunk 2]
```diff
@@ -251,30 +277,35 @@ func processSubscription(req *push.ChannelReq, config *configType) {
...
 
-func handlePushResponse(batch *batchResponse, messages []fcm.MessageData) {
+func handlePushResponse(batch *batchResponse, messages []*fcmv1.Message, uids []types.Uid) {
 	if batch.FailureCount <= 0 {
 		return
 	}
 
 	for i, resp := range batch.Responses {
+
+		logs.Warn.Println("tnpg error:", resp.ErrorCode, resp.ErrorMessage, resp.ExtendedError, resp.Code)
```
Comment (aforge) [inline server/push/tnpg/push_tnpg.go:287]: remove?

**Comment by aforge** [comment]
> But voip pushes arrive as normal pushes as opposite to PushKit despite the fact that they are sent as voip.

Humm... that sounds like FCM won't actually send PushKit/voip pushes?
It defeats the purpose of this change...

File: server/push/tnpg/push_tnpg.go (hunk 2 above)
Comment (or-else) [reply server/push/tnpg/push_tnpg.go:287]: Removed.

File: server/push/fcm/payload.go (hunk 1 above)
Comment (or-else) [reply server/push/fcm/payload.go:205]: It should although the difference does not affect anything. Fixed anyway.

**Comment by or-else** [comment]
> It defeats the purpose of this change...

Some people say it should work, some that it won't. FCM returns no error when I send voip pushes. I'll try to debug it first. Adding APNS is a lot of work.

**Comment by or-else** [comment]
Alternatively maybe we can use something like https://github.com/appleboy/gorush

**Review by aforge — APPROVED** [review]

**Pull request merged**